public/app/core/components/Login/LoginCtrl.tsx

```diff
@@ -0,0 +1,141 @@
+import React from 'react';
+import appEvents from 'app/core/app_events';
+import config from 'app/core/config';
+
+import { updateLocation } from 'app/core/actions';
+import { connect } from 'react-redux';
+import { StoreState } from 'app/types';
+import { PureComponent } from 'react';
+import { getBackendSrv } from '@grafana/runtime';
+import { hot } from 'react-hot-loader';
+
+export interface FormModel {
+  user: string;
+  password: string;
+  email: string;
+}
+interface Props {
+  routeParams?: any;
+  updateLocation?: typeof updateLocation;
+  children: (obj: {
+    isLoggingIn: boolean;
+    changePassword: (pw: { newPassword: string; confirmNew: string; oldPassword: string }, valid: boolean) => void;
+    isChangingPassword: boolean;
+    toGrafana: Function;
+    login: (data: FormModel, valid: boolean) => void;
+  }) => JSX.Element;
+}
+
+interface State {
+  isLoggingIn: boolean;
+  isChangingPassword: boolean;
+}
+
+export class LoginCtrl extends PureComponent<Props, State> {
+  result: any = {};
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      isLoggingIn: false,
+      isChangingPassword: false,
+    };
+    if (config.loginError) {
+      appEvents.on('alert-warning', ['Login Failed', config.loginError]);
+    }
+  }
+
+  changePassword = (pws: { newPassword: string; confirmNew: string }, valid: boolean) => {
+    console.log('Changing PW');
+    if (!valid) {
+      return;
+    }
+
+    getBackendSrv()
+      .put('/api/user/password', pws)
+      .then(() => {
+        this.toGrafana();
+      })
+      .catch((err: any) => console.log(err));
+  };
+
+  login = (formModel: FormModel, valid: boolean) => {
+    console.log(formModel);
+    console.log(this);
+    if (!valid) {
```

Comment:
Should be handled in the form itself, `login` should only control the login action. that should not be called at all if the form is not valid